sqlx-mysql/src/connection/compression.rs

```diff
@@ -0,0 +1,214 @@
+use crate::protocol::Capabilities;
+#[cfg(feature = "compression")]
+use crate::Compression;
+use crate::CompressionConfig;
+#[cfg(feature = "compression")]
+use compressed_stream::CompressedStream;
+use sqlx_core::io::{ProtocolDecode, ProtocolEncode};
+use sqlx_core::net::{BufferedSocket, Socket};
+use sqlx_core::Error;
+
+pub(crate) struct CompressionMySqlStream<S = Box<dyn Socket>> {
+    stream: CompressionStream,
+    pub(crate) socket: BufferedSocket<S>,
+}
+
+impl<S: Socket> CompressionMySqlStream<S> {
+    pub(crate) fn not_compressed(socket: BufferedSocket<S>) -> Self {
+        let stream = CompressionStream::NotCompressed;
+        Self { stream, socket }
+    }
+
+    #[cfg(feature = "compression")]
+    fn compressed(socket: BufferedSocket<S>, compression: CompressionConfig) -> Self {
+        let stream = CompressionStream::Compressed(CompressedStream::new(compression));
+        Self { stream, socket }
+    }
+
+    pub(crate) fn create(
+        socket: BufferedSocket<S>,
+        #[cfg_attr(not(feature = "compression"), allow(unused_variables))]
+        capabilities: &Capabilities,
+        compression: Option<CompressionConfig>,
+    ) -> Self {
+        match compression {
+            #[cfg(feature = "compression")]
+            Some(c) if c.is_supported(&capabilities) => {
```

Comment:
It could be helpful to emit a `debug` log here if the user enabled compression but the server didn't support it.